scripts/ndjson_to_csv/ndjson_to_csv.py

```diff
@@ -0,0 +1,122 @@
+# /usr/bin/python3
+
+import csv, json, sys, tempfile
+
+from functools import reduce
+from pathlib import Path
+
+import pandas as pd
+from tqdm import tqdm
+
+
+col_headers = set()
+help_message = 'Convert FHIR data stored in an NDJSON file to a CSV format.\n\
+\n\
+Usage:\n\
+\n\
+    ndjson_to_csv.py <inputfile> <output_filename>\n\
+\n\
+\n\
+<inputfile>        Source file to be converted (.json, .ndjson, .txt)\n\
+<output_filename>  Optional desired output filename\n\
+\n\
+-h, --help         view help message\n'
+
+def flatten(data, parent_key='', sep='.'):
+    items = []
+    for k, v in data.items():
+        new_key = parent_key + sep + k if parent_key else k
+        if isinstance(v, dict):
+            items.extend(flatten(v, new_key, sep).items())
+        elif isinstance(v, list):
+            if not isinstance(v[0], (str, int)):
+                if len(v) > 1:
+                    for index, item in zip(range(len(v)), v):
+                        if index > 0:
+                            items.extend(flatten(
+                                item,
+                                '{}{}'.format(new_key, index+1),
+                                sep
+                            ).items())
+                        else:
+                            items.extend(flatten(item, new_key, sep).items())
+                else:
+                    for item in v:
+                        items.extend(flatten(item, new_key, sep).items())
+            else:
+                list_strings = [str(item) for item in v]
+                col_headers.add(new_key)
+                items.append((new_key, ' '.join(list_strings)))
+        else:
+            col_headers.add(new_key)
+            items.append((new_key, v))
+    return dict(items)
+
+
+if __name__=='__main__':
+    """
+    Convert FHIR data stored in an NDJSON file to a CSV format.
+
+    Usage: ndjson_to_csv.py <inputfile> <output_filename>
+    
+    <inputfile>        Source file to be converted (.json, .ndjson, .txt)
+    <output_filename>  Optional desired output filename
+    """
+
+    argv = sys.argv[1:]
+    
+    if '-h' in argv or '--help' in argv:
+        print(help_message)
+        exit()
+    if len(argv) > 2:
+        print('Too many arguments.')
+        print(help_message)
+        exit()
+
+    inputfile, outputfile = argv[0], argv[1]
+    
+    if outputfile[-4:] != '.csv':
+        outputfile = f'{outputfile}.csv'
+
+    # Error handling
+    if not Path(inputfile).is_file():
+        print(f'Source file "{inputfile}" does not exist.')
+        exit()
+    if Path(outputfile).is_file():
+        print(f'A file with the name "{outputfile}" already exists in this directory.')
+        exit()
+
+    # Create Tempfile
+    tf = tempfile.NamedTemporaryFile(prefix=f'{outputfile}_temp_', mode="w+t", delete=True)
```

Comment:
This may be fine, but you also may need to benchmark this against some example NDJSON files to test performance....

First, great thought on having to iterate through the file twice to accurately get a list of the CSV headers.  I see that you're writing the data to a temp file to avoid having to call flatten twice per line, however there is a cost there in you'll need double the amount of file space to render the result (1 output file for the temp file and another for the csvwriter).  The first should get deleted when the program exits, but that won't be to the very end.

It be interesting to see what this looks like if the first pass only adds to the `col_header` set, and the second pass writes the data to file.  It may be slower computationally, but I believe it would require fewer resources to compute.

Just a thought.

Reply:
> It be interesting to see what this looks like if the first pass only adds to the col_header set, and the second pass writes the data to file. It may be slower computationally, but I believe it would require fewer resources to compute.

I'm pretty sure the most time consuming part of this is building the CSV. In the interest of saving disk space, this might be worth it.

Reply:
TL;DR - I'm pushing the changes you suggested and we're flattening the file x2

OK, so I created a larger test file ~340 MB to try and get a better idea for how performant these 2 options are.

**Option 1 (Tempfile)** - Completed in ~47 seconds
**Option 2 (Flatten 2x)** - Completed in ~55 seconds

So by using the tempfile approach we save about 8 seconds and for a brief period triple the disk space used. That extrapolates to 23.5 sec / 1 GB. While that will add up quickly for people who have larger amounts of data, I'm sure they'll appreciate not creating unnecessarily large files on their systems.

If you're curious, it looks as though each pass through `flatten()` was taking ~12 seconds.